src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseFVMKernels.hpp

```diff
@@ -850,6 +854,721 @@ class FaceBasedAssemblyKernelFactory
+        real64 dDens_dC[numComp]{};
+
+        real64 const trans[numFluxSupportPoints] = { stack.transmissibility[connectionIndex][0],
+                                                     stack.transmissibility[connectionIndex][1] };
+
+        //***** calculation of flux *****
+        // loop over phases, compute and upwind phase flux and sum contributions to each component's flux
+        for( integer ip = 0; ip < m_numPhases; ++ip )
+        {
+
+          // loop over components
+          for( integer ic = 0; ic < numComp; ++ic )
+          {
+
+            real64 compFracGrad = 0.0;
+            real64 dCompFracGrad_dP[numFluxSupportPoints]{};
+            real64 dCompFracGrad_dC[numFluxSupportPoints][numComp]{};
+
+            // compute the component fraction gradient using the diffusion transmissibility
+            computeCompFractionGradient( ip, ic,
+                                         seri, sesri, sei,
+                                         trans,
+                                         compFracGrad,
+                                         dCompFracGrad_dP,
+                                         dCompFracGrad_dC );
+
+            // choose upstream cell for composition upwinding
+            localIndex const k_up = (compFracGrad >= 0) ? 0 : 1;
+
+            localIndex const er_up  = seri[k_up];
+            localIndex const esr_up = sesri[k_up];
+            localIndex const ei_up  = sei[k_up];
+
+            // computation of the upwinded mass flux
+            real64 const upwindCoefficient =
+              m_referencePorosity[er_up][esr_up][ei_up] *
+              m_phaseDiffusivityMultiplier[er_up][esr_up][ei_up][0][ip] *
+              m_phaseDens[er_up][esr_up][ei_up][0][ip] * m_phaseVolFrac[er_up][esr_up][ei_up][ip];
+            diffusionFlux[ic] += upwindCoefficient * compFracGrad;
+
+            // add contributions of the derivatives of component fractions wrt pressure/component fractions
+            for( integer ke = 0; ke < numFluxSupportPoints; ke++ )
+            {
+              dDiffusionFlux_dP[ke][ic] += upwindCoefficient * dCompFracGrad_dP[ke];
+              for( integer jc = 0; jc < numComp; ++jc )
+              {
+                dDiffusionFlux_dC[ke][ic][jc] += upwindCoefficient * dCompFracGrad_dC[ke][jc];
+              }
+            }
+
+            // add contributions of the derivatives of upwind coefficient wrt pressure/component fractions
+            real64 const dUpwindCoefficient_dP =
+              m_referencePorosity[er_up][esr_up][ei_up] *
+              m_phaseDiffusivityMultiplier[er_up][esr_up][ei_up][0][ip] *
+              ( m_dPhaseDens[er_up][esr_up][ei_up][0][ip][Deriv::dP] * m_phaseVolFrac[er_up][esr_up][ei_up][ip]
+                + m_phaseDens[er_up][esr_up][ei_up][0][ip] * m_dPhaseVolFrac[er_up][esr_up][ei_up][ip][Deriv::dP] );
+            dDiffusionFlux_dP[k_up][ic] += dUpwindCoefficient_dP * compFracGrad;
+
+            applyChainRule( numComp,
```

Comment:
sorry, i don't understand that applyChainRule here, why can't we use m_dPhaseDens[er_up][esr_up][ei_up][0][ip][Deriv::dC + jc] directly below?

Reply:
We need to use the chain rule here because m_dPhaseDens[er_up][esr_up][ei_up][0][ip][Deriv::dC + jc] is computed by the fluid model as the derivative of phase density (rho_l) with respect to component fractions (z_c), and we need to assemble derivatives of the mass conservation equation with respect to component densities (rho_c).

So, we have to first compute (drho_l/drho_c) = (drho_l/dz_c)*(dz_c/drho_c) and then use the result to compute the derivative of the diffusive flux with respect to component densities.

Reply:
why don't we do it right after fluid model compute once and for all? it's really confusing...

Reply:
Yes, we can definitely do it as another state update step after calling the fluid model. There is quite a bit of code to modify, but it would simplify a lot of things (and possibly avoid future implementation mistakes)

Reply:
As designed, fluid model owns its output quantities and derivatives, and always computes the derivatives w.r.t. the input variables of the fluid model interface, i.e. `z_c`. Thus the model is independent of the specific variable formulation implemented in a solver (as there could be multiple solvers using the same fluid model). Applying the chain rule in-place (i.e. overwriting `drho_l/dz_c` with `drho_l/drho_c` in memory) would be seen as an anti-pattern in this design as the solver shouldn't overwrite data it doesn't "own", while allocating and storing a separate copy of `drho_l/drho_c` would be prohibitively expensive in terms of memory footprint (once this is done for all quantities). Hence we apply the chain rule on the fly where needed, but suggestions are welcome.

Reply:
I am talking about solver code outside of fluid model. My expectation is that derivatives will be consistent for all quantities and there is no need to remember that specifically for this one we need to chain rule.